### PR TITLE
Add all fields for `Tool` struct

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

This exhaustively adds all fields to the `Tool` struct and the API around it

## 🚫 Blocked by

- #2243 

## 🔍 What does this change?

- Add `PropertyBag`
- Add fields to `Tools`
- Greatly extend documentation and tests

## 🛡 What tests cover this?

- This branch has 💯% test coverage

### CODE REVIEW

File: libs/sarif/src/schema/run.rs [hunk 1]
```diff
@@ -22,6 +22,16 @@ pub struct Run {
 
 impl Run {
     /// Create a new `Run` with the given tool.
+    ///
```
Comment (indietyp) [inline libs/sarif/src/schema/run.rs:25]: Run as a missing property bag. (see note about `3.8`)

File: libs/sarif/src/schema/tool.rs [hunk 2]
```diff
@@ -39,6 +153,16 @@ pub struct ToolComponent {
 
 impl ToolComponent {
     /// Create a new `ToolComponent` with the given name.
+    ///
```
Comment (indietyp) [inline libs/sarif/src/schema/tool.rs:156]: Run as a missing property bag. (see note about `3.8`)

**Review by indietyp — CHANGES_REQUESTED** [review]
looks good overall

File: libs/sarif/src/schema/properties.rs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    pub tags: BTreeSet<Cow<'static, str>>,
+
+    /// A dictionary, each of whose keys specifies a distinct additional information element and
+    /// each of whose values provides the information.
+    #[cfg(feature = "serde")]
+    #[serde(flatten)]
+    pub additional: BTreeMap<Cow<'static, str>, serde_json::Value>,
+}
+
+impl PropertyBag {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply libs/sarif/src/schema/properties.rs:27]: I marked all types, which are not fully implemented yet, as non-exhaustive, so I'll add `PropertyBag` to them later.

File: libs/sarif/src/schema/tool.rs [hunk 4]
```diff
@@ -1,31 +1,145 @@
...
+    /// ]);
+    ///
+    /// assert_eq!(tool.extensions, vec![
+    ///     ToolComponent::new("prettier-plugin-packagejson"),
+    ///     ToolComponent::new("prettier-plugin-sh"),
+    ///     ToolComponent::new("prettier-plugin-sql"),
+    /// ]);
+    /// ```
+    #[must_use]
+    pub fn with_extensions(mut self, extension: impl IntoIterator<Item = ToolComponent>) -> Self {
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/sarif/src/schema/tool.rs:101]: consistent w/ #2243, therefore fine keeping it

**Review by indietyp — APPROVED** [review]

**Review by indietyp — APPROVED** [review]

**Pull request merged**